From 0764425f7301f1d792f6c06be1d26467a7915af5 Mon Sep 17 00:00:00 2001
From: Alessandro Astone <alessandro.astone@canonical.com>
Date: Wed, 30 Sep 2026 13:06:01 +0200
Subject: [PATCH] settings-impl-portal: Read non-standard settings from snaps
 too

And add an in-line comment explaining why we may skip parsing these settings.

Forwarded: https://gitlab.gnome.org/GNOME/libadwaita/-/merge_requests/1831
---
 src/adw-settings-impl-portal.c | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/src/adw-settings-impl-portal.c b/src/adw-settings-impl-portal.c
index 809e7b83..b60d6b38 100644
--- a/src/adw-settings-impl-portal.c
+++ b/src/adw-settings-impl-portal.c
@@ -248,6 +248,19 @@ is_running_in_flatpak (void)
 #endif
 }
 
+static gboolean
+is_running_in_snap (void)
+{
+  return g_getenv ("SNAP") != NULL;
+}
+
+static gboolean
+may_have_gsettings (void)
+{
+  return !is_running_in_flatpak () &&
+         !is_running_in_snap (); // TODO: snaps could be unsandboxed, or have a sandbox hole for gsettings
+}
+
 AdwSettingsImpl *
 adw_settings_impl_portal_new (gboolean enable_color_scheme,
                               gboolean enable_high_contrast,
@@ -319,7 +332,8 @@ adw_settings_impl_portal_new (gboolean enable_color_scheme,
     g_variant_unref (variant);
   }
 
-  if (is_running_in_flatpak ()) {
+  // non-standard portal settings; only read from the portal if there's no better fallback
+  if (!may_have_gsettings ()) {
     if (enable_document_font_name &&
         read_setting (self, "org.gnome.desktop.interface",
                       "document-font-name", "s", &variant)) {
-- 
2.53.0

